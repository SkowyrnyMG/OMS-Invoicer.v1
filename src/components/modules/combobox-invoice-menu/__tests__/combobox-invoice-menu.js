import React from 'react';

import {
  renderWithReduxRouter,
  screen,
  fireEvent,
  leftClick,
} from 'utils/tests/test-helper';
import ComboboxInvoiceMenu from '../combobox-invoice-menu';

const testItems = [
  {
    customer_vat: 'customer-test-vat',
    customer_name: 'customer-test-name',
    order_number: 'test_order_number',
    price: 500,
    currency: 'EUR',
  },
];

const testSearchResult = 'test_order_number - 500 EUR - customer-test-name';

const testHandleSetItemFn = jest.fn(() =>
  console.log('Values succesfully set to the inputs.'),
);
const testHandleResetItemFn = jest.fn(() =>
  console.log('Values have been reseted!'),
);

describe('ComboboxInvoiceMenu', () => {
  test('should display in the document with correct styles', () => {
    renderWithReduxRouter(
      <ComboboxInvoiceMenu
        items={testItems}
        handleSetItemFn={testHandleSetItemFn}
        handleResetItemFn={testHandleResetItemFn}
      />,
    );

    const ComboInvoiceMenuComponent = screen.getByTestId(
      'combobox-invoice-menu',
    );

    expect(ComboInvoiceMenuComponent).toBeInTheDocument();
    expect(ComboInvoiceMenuComponent).toMatchSnapshot();
  });

  test('should not display combo list when search input is untouched', () => {
    renderWithReduxRouter(
      <ComboboxInvoiceMenu
        items={testItems}
        handleSetItemFn={testHandleSetItemFn}
        handleResetItemFn={testHandleResetItemFn}
      />,
    );

    const comboList = screen.queryByRole('listbox', {
      name: 'Choose a finished order:',
    });

    expect(comboList).toBeNull();
  });

  test('should display combo list after sucesfull match with search query string and store user choice in combo input', () => {
    renderWithReduxRouter(
      <ComboboxInvoiceMenu
        items={testItems}
        handleSetItemFn={testHandleSetItemFn}
        handleResetItemFn={testHandleResetItemFn}
      />,
    );

    // * type search string into search input, after succesful match comboList should be displayed
    let comboInput = screen.getByPlaceholderText(
      /provide order number or company name/i,
    );

    fireEvent.change(comboInput, { target: { value: 'customer-test' } });

    let comboList = screen.queryByRole('listbox', {
      name: 'Choose a finished order:',
    });

    expect(comboList).toBeInTheDocument();
    expect(comboList.children).not.toBeNull();

    const comboListItem = screen.getByRole('option', {
      name: testSearchResult,
    });

    fireEvent.click(comboListItem, leftClick);

    // * after successful option click combolist should dissapear and combo input should receive option value
    comboList = screen.queryByRole('listbox', {
      name: 'Choose a finished order:',
    });
    comboInput = screen.getByPlaceholderText(
      /provide order number or company name/i,
    );

    expect(comboList).toBeNull();
    expect(comboInput.value).toEqual(testSearchResult);
  });

  test('should reset comboInput after click on the reset button', () => {
    renderWithReduxRouter(
      <ComboboxInvoiceMenu
        items={testItems}
        handleSetItemFn={testHandleSetItemFn}
        handleResetItemFn={testHandleResetItemFn}
      />,
    );

    const comboInput = screen.getByPlaceholderText(
      /provide order number or company name/i,
    );
    const resetButton = screen.getByRole('button', { name: /reset/i });

    fireEvent.change(comboInput, { target: { value: testSearchResult } });
    expect(comboInput.value).toEqual(testSearchResult);

    fireEvent.click(resetButton, leftClick);
    expect(comboInput.value).toBeFalsy();
    expect(testHandleResetItemFn).toHaveBeenCalledTimes(1);
  });

  test('should use handleSetItemFn after clicking accept button', () => {
    renderWithReduxRouter(
      <ComboboxInvoiceMenu
        items={testItems}
        handleSetItemFn={testHandleSetItemFn}
        handleResetItemFn={testHandleResetItemFn}
      />,
    );
    const acceptButton = screen.getByRole('button', { name: /accept/i });

    fireEvent.click(acceptButton, leftClick);

    expect(testHandleSetItemFn).toHaveBeenCalledTimes(1);
  });

  test('should open and close comboList on toggle button click', () => {
    renderWithReduxRouter(
      <ComboboxInvoiceMenu
        items={testItems}
        handleSetItemFn={testHandleSetItemFn}
        handleResetItemFn={testHandleResetItemFn}
      />,
    );

    const toggleButton = screen.getByRole('button', { name: /toggle menu/i });
    let comboList = screen.queryByRole('listbox', {
      name: 'Choose a finished order:',
    });

    expect(comboList).toBeNull();

    fireEvent.click(toggleButton, leftClick);

    comboList = screen.queryByRole('listbox', {
      name: 'Choose a finished order:',
    });
    expect(comboList).toBeInTheDocument();

    fireEvent.click(toggleButton, leftClick);
    comboList = screen.queryByRole('listbox', {
      name: 'Choose a finished order:',
    });
    expect(comboList).toBeNull();
  });
});
