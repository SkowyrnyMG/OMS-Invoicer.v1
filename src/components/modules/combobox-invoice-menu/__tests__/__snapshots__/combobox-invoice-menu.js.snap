// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboboxInvoiceMenu should display in the document with correct styles 1`] = `
.c3 {
  padding: 0.5rem 1rem;
  border: 2px solid #446DF6;
  border-radius: 0.5rem;
  background: none;
  cursor: pointer;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

.c3:not(:first-child) {
  margin-left: 2rem;
}

.c3:hover {
  background: #446DF6;
  color: #FFFFFF;
}

.c4 {
  position: absolute;
  max-height: 25rem;
  padding: 2rem;
  list-style: none;
  overflow-y: auto;
  background: #F8F8FF;
  box-shadow: 0 4px 5px rgba(0,0,0,0.1);
  opacity: 0;
  visibility: hidden;
  z-index: 1000;
}

.c1 {
  padding-left: 0.5rem;
  margin-bottom: 1rem;
  width: 100%;
  height: 3rem;
  font-size: 16px;
}

.c0 {
  grid-column: 1 / -1 !important;
  position: relative;
  margin-bottom: 3rem;
}

.c5 {
  cursor: pointer;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c6 {
  opacity: 0;
  display: none;
}

<div
  class="c0"
  data-testid="combobox-invoice-menu"
>
  <label
    for="downshift-0-input"
    id="downshift-0-label"
  >
    Choose a finished order:
  </label>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="downshift-0-menu"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-0-menu"
      aria-labelledby="downshift-0-label"
      autocomplete="off"
      class="c1"
      id="downshift-0-input"
      placeholder="Provide order number or company name"
      value=""
    />
    <div
      class="c2"
    >
      <button
        aria-label="toggle menu"
        class="c3"
        id="downshift-0-toggle-button"
        tabindex="-1"
        type="button"
      >
        Show
      </button>
      <button
        class="c3"
        type="button"
      >
        Accept
      </button>
      <button
        class="c3"
        type="button"
      >
        Reset
      </button>
    </div>
  </div>
  <ul
    aria-labelledby="downshift-0-label"
    class="c4"
    id="downshift-0-menu"
    role="listbox"
  >
    <li
      aria-selected="false"
      class="c5"
      id="downshift-0-item-0"
      role="option"
    >
      test_order_number - 500 EUR - customer-test-name
    </li>
    <div
      class="c6"
    >
      Nothing found...
    </div>
  </ul>
</div>
`;
